### PR TITLE
[Remote Tab Management] Receive Close Tab commands from other devices

### PR DESCRIPTION
**Pull request by skhamis**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-9398)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/20808)

First of the Remote Tab Management PRs. Couple of things here:
- Add Nimbus feature flag to be able to control it while developing on it
- Added a secret menu option
- Renamed SendTabDelegate to FxACommandsDelegate to encompass CloseTab command and future ones
- Upon receiving a "CloseTab" command from FxA, go through all tabs and close the first matching Url

## :pencil: Checklist
You have to check all boxes before merging
- [ ] Filled in the above information (tickets numbers and description of your work)
- [ ] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed, I updated documentation / comments for complex code and public methods
- [ ] If needed, added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: firefox-ios/Storage/Rust/RustRemoteTabs.swift
Comment (skhamis) [inline firefox-ios/Storage/Rust/RustRemoteTabs.swift:17]: This commit is stacked on my TabsStore Refactor, which will land before this: https://github.com/mozilla-mobile/firefox-ios/pull/20776

File: firefox-ios/Client/ApplicationHelper.swift
Comment (skhamis) [inline firefox-ios/Client/ApplicationHelper.swift:55]: Most of the heavy lifting here -- not sure if this is the best way, but I'll let a FxIOS dev chime in here!

File: firefox-ios/Client/ApplicationHelper.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/ApplicationHelper.swift:53]: This line shouldn't be necessary, AFAICS? We should be able to simply iterate `allWindowTabManagers().forEach` I believe.

File: firefox-ios/Client/ApplicationHelper.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/ApplicationHelper.swift:56]: Ideally I don't think we should be typecasting to `TabManagerImplementation` here. I would suggest that we add the new `closeTab` function to the `TabManager` protocol (though might be good to get additional eyes on this from one of our tab experts cc @OrlaM @nbhasin2 ).

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/TabManagement/TabManagerImplementation.swift:476]: This is not really specific to this PR, but it seems somewhat fragile for us to be closing tabs by URL. There are various scenarios in which multiple different tabs might have the same URL but with different content based on how the user has interacted with the webpage. Ideally I'd think we'd want to identify tabs by a UUID rather than the URL. I'm guessing this is part of backwards-compatibility with older code though and is probably not something we could change easily.

**Review by mattreaganmozilla — CHANGES_REQUESTED** [review]
A couple small comments: can we possibly avoid the typecast and scene iteration? I don't believe those should be necessary.

File: firefox-ios/Client/TabManagement/TabManagerImplementation.swift
Comment (skhamis) [reply firefox-ios/Client/TabManagement/TabManagerImplementation.swift:476]: Great points! The thing here is we are closing tabs remotely from _other_ devices, synced tabs doesn't keep track of any platforms specific ids and we initially thought of using some kind've id in the payloads -- but that adds a whole nest of other things like, if a user navigates to a new page on the same tab -- do we still close it? etc).

However maybe I should rename this method to be explicitly clear its for use of closing remote tabs.

File: firefox-ios/Client/ApplicationHelper.swift
Comment (skhamis) [reply firefox-ios/Client/ApplicationHelper.swift]: Great suggestion here -- removed all the unnecessary casting and simplified the call-sites. Hopefully looks cleaner now!

File: firefox-ios/Client/Application/AppDelegate+SyncSentTabs.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/Application/AppDelegate+SyncSentTabs.swift:58]: Since this async call is just calling a `Task` which wraps everything else I don't believe it's necessary (the `mainQueue.async` won't have any effect AFAICS), the Task will be executed on a background queue. LMK if I'm missing something here however.

```suggestion
            Task {
                for url in urls {
                    await self.applicationHelper.closeTab(url)
                }
            }
```

File: firefox-ios/Client/ApplicationHelper.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/ApplicationHelper.swift:62]: This seems like it could be cleaned up slightly but LMK if I'm misunderstanding. Possibly something like:

```
windowManager.allWindowTabManagers().forEach {
    await _ = $0.removeTabByUrl(url)
}
```

Comment (skhamis) [reply firefox-ios/Client/ApplicationHelper.swift:62]: ah my goal was to exit out of the loop once we find a matching url so we don't unnecessarily iterate through all the tabs and end up removing more than one instance of the url (if there were duplicates).

File: firefox-ios/Client/Application/AppDelegate+SyncSentTabs.swift
Comment (skhamis) [reply firefox-ios/Client/Application/AppDelegate+SyncSentTabs.swift]: good catch! I was mainly copying how send tabs does it above this one and figured there was a "good" reason. However I think you're right it's unnecessary in this scenario -- everything seems to work okay in testing!

File: firefox-ios/Client/ApplicationHelper.swift
Comment (mattreaganmozilla) [reply firefox-ios/Client/ApplicationHelper.swift:62]: Ah ok sorry I misunderstood the intent 👍

File: firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: cc @cyndichin Since this might have some overlap with the proposed debug feature flag screen.

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift
Comment (mattreaganmozilla) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:636]: Minor nit: I think you could just use `normalTabs` here.

File: firefox-ios/nimbus-features/remoteTabManagement.yaml
Comment (mattreaganmozilla) [inline firefox-ios/nimbus-features/remoteTabManagement.yaml:2]: I believe there is a specific PR review process for new Nimbus features but @adudenamedruby could confirm for certain.

File: firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: thanks for tagging me @mattreaganmozilla! Ideally if this is purely for testing it can live within the feature flag section. However, I do not want that proposed screen to be a blocker for this PR and can refactor at a later date if warranted. 

@skhamis do you mind providing more context on the purpose of this setting (i.e. purely testing) and why we are using the different channels `.buildAndUser` and `.userOnly` for the `.remoteTabManagement` feature flag? Thanks!

Comment (skhamis) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: Hi @cyndichin! This feature will be released on Android & Desktop soon and overall the feature is still relatively new. The goal(s) for this setting is:
1. Control whether we broadcast the capability from iOS to other platforms
    - Ensures we don't unnecessarily launch this feature too early on iOS 
2. Eventually control whether we show the user that they can close tabs on other devices on iOS (this will be in a separate PR)
3. Ability to use nimbus to shutdown the feature if something goes wrong down the line 

Happy to make changes or use the new menus if they support the above cases!!

File: firefox-ios/nimbus-features/remoteTabManagement.yaml
Comment (adudenamedruby) [reply firefox-ios/nimbus-features/remoteTabManagement.yaml:2]: Nope, no fancy process for Nimbus; this looks fine to me.
You might be thinking of adding telemetry, where you'd need a data review. :)

**Review by mattreaganmozilla — APPROVED** [review]
Barring any follow-up comments from @linabutler or @cyndichin, changes lgtm.

**Comment by linabutler** [comment]
Thanks, @skhamis and @mattreaganmozilla! I'll set aside some time in the next couple of days to review, and then we can get this landed!

File: firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: @skhamis I have a draft for the PR, but the new screen is not ready yet. I will probably get to it later this week. Do you think we can hold off on adding this hidden setting until then? 

Otherwise, I would recommend we wrap this code for debug since our secret settings can be accessed in production and based on your notes it seems that we don't want users to access it. Also, the current capability is that nimbus is not necessarily shutting down the feature, this is changing the user default setting on a specific device.

Thanks! Let me know if you need more clarity.

Comment (skhamis) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: Definitely don't mind waiting to get this landed till next week if needed. This feature will have multi-parts so there's no rush landing this (though it'd be nice).

> Otherwise, I would recommend we wrap this code for debug since our secret settings can be accessed in production

I don't think we're necessarily concerned with users accessing it, the feature will work, it just won't have some of the polish we'll eventually be adding (notifications, etc). This is also a receive-only capability, however if we're still better off waiting I'll defer to the iOS team on this!

Comment (cyndichin) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: Okay, if no concerns in exposing the feature, then no issue on my side in regards to the PR! 

We have some time before v130 gets released so can follow up with you on this when the debug feature flags PR gets merged. However, that new screen will be only available in beta or developer build. Are there concerns if we hide this work from users eventually?

Comment (skhamis) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: I don't think there's any concern to hide this, as long as developer/beta builds are able to toggle it (which sounds like it will be) because eventually we'll just flip the nimbus variable to be default on when we want to enable the feature. 

Appreciate looking into this and helping us out navigate the right away to go about feature flagging!

File: firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift
Comment (linabutler) [inline firefox-ios/Client/TabManagement/Legacy/LegacyTabManager.swift:635]:
```suggestion
    func removeTabs(with urls: [URL]) async -> Bool {
```

Thought: How would you feel about changing this signature to take multiple URLs (and then changing `ApplicationHelper.closeTab` to `closeTabs`?)

That way, we can turn `urls` into a `Set`, filter `normalTabs` to find the tabs for all the URLs in the set in just one pass, then remove each one. It would make this an `O(tabs.count)` operation, rather than `O(tabs.length * urls.count)`.

Something like:

```swift
func removeTabs(with urls: [URL]) async {
    let urls = Set(urls)
    let tabsToRemove = normalTabs.filter { tab in
        guard let url = tab.url else { return false }
        return urls.contains(url)
    }
    for tab in tabs {
        await withCheckedContinuation { continuation in
            removeTab(tab) { continuation.resume() }
        }
    }
}
```

(As a bonus, we could also optimize `removeTab(_:completion:)` to update the selected index just once, but one step at a time!)

File: firefox-ios/Client/Application/AppDelegate+SyncSentTabs.swift
Comment (linabutler) [inline firefox-ios/Client/Application/AppDelegate+SyncSentTabs.swift:53]: (I left a comment below about passing the list all the way down to the `TabManager`, so that it can remove them all at once).

File: firefox-ios/Client/FeatureFlags/NimbusFlaggableFeature.swift
Comment (linabutler) [inline firefox-ios/Client/FeatureFlags/NimbusFlaggableFeature.swift:34]:
```suggestion
    case closeRemoteTabs
```

Nitpick: How would you feel about changing this flag's name to `closeRemoteTabs`? `remote-tab-management` is the name of the Nimbus feature; `close-tabs` is the actual variable.

If we introduce other variables in the `remote-tab-management` feature (`close-all-tabs`, maybe? `move-tab`? 😅), it might be a little confusing to have the `remoteTabManagement` flag actually just mean "close tabs".

File: firefox-ios/Providers/Profile.swift
Comment (linabutler) [inline firefox-ios/Providers/Profile.swift:611]: Issue: Please keep the `default` case; that'll avoid breaking the iOS build when we add new commands.

(We unfortunately can't mark `IncomingDeviceCommand` as non-frozen to have the compiler enforce that, because I think the only way to mark enums as non-frozen is to build `MozillaRustComponentsSwift` with [library evolution](https://www.swift.org/blog/library-evolution/) enabled).

File: firefox-ios/Providers/Profile.swift
Comment (linabutler) [inline firefox-ios/Providers/Profile.swift:613]: Nice cleanup! 🏆

File: firefox-ios/RustFxA/RustFirefoxAccounts.swift
Comment (linabutler) [inline firefox-ios/RustFxA/RustFirefoxAccounts.swift:169]: Issue / Question: Hmmm, will this do the right thing if the user hasn't flipped the secret setting on? That is, if the feature is enabled via the Nimbus variable, and the user hasn't ever opened secret settings, will the pref ever be set?

I wonder if this module can export a new option set, like:

```swift
public struct RustFirefoxAccountsFeatures: OptionSet {
    let rawValue: Int

    static let closeRemoteTabs = RustFirefoxAccountsFeatures(rawValue: 1 << 0)
}
```

Then have `RustFirefoxAccounts.startup()` take that option set, and, finally, have the Firefox for iOS `Client` code get the Nimbus feature variable, set up the option set, and pass it to `startup`.

**Review by linabutler — CHANGES_REQUESTED** [review]
Nice work, @skhamis! I wouldn't mind looking at this again, but I didn't have many comments, and I think this is just about good to go! 🎉

File: firefox-ios/RustFxA/RustFirefoxAccounts.swift
Comment (skhamis) [reply firefox-ios/RustFxA/RustFirefoxAccounts.swift]: Great catch! I think your correct in that the nimbus flag won't be used (only the user-specified.

I love the OptionSet Idea! There are a few places startup (and reconfig) are called that makes it slightly hard as it made a spiral effect of trying to pull feature flags everywhere. However I think we only care about AppDelegate right now so I think it's in a much better state.

File: firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift:7]: Btw, the debug setting menu has been merged, if you would like to add your toggle there in this PR or a separate PR.

Comment (skhamis) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift]: @cyndichin I have updated it to use the new debug setting (I made it a separate commit so it's easier to review) -- seemed fairly straightforward and everything testing okay, but let me know if I missed anything!

File: firefox-ios/Client/Frontend/Settings/Main/Debug/FeatureFlags/FeatureFlagsDebugViewController.swift
Comment (cyndichin) [inline firefox-ios/Client/Frontend/Settings/Main/Debug/FeatureFlags/FeatureFlagsDebugViewController.swift:66]: @skhamis we don't need to set the user defaults here, I had to do it for the microsurvey to reset the expiration that prevented the survey to show more than once as part of the messaging system. 

If there's no additional action needed for the close remote tab setting to trigger it to show / hide, then you don't need to add this.

File: firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift
Comment (cyndichin) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/ToggleCloseRemoteTabs.swift]: thanks! I left one comment that should be addressed, but otherwise looks great. thank you for making a separate commit :)

File: firefox-ios/Client/Frontend/Settings/Main/Debug/FeatureFlags/FeatureFlagsDebugViewController.swift
Comment (skhamis) [reply firefox-ios/Client/Frontend/Settings/Main/Debug/FeatureFlags/FeatureFlagsDebugViewController.swift]: ah gotcha, sounds good -- updated!

File: firefox-ios/Client/FeatureFlags/NimbusFlaggableFeature.swift
Comment (skhamis) [reply firefox-ios/Client/FeatureFlags/NimbusFlaggableFeature.swift]: That definitely makes more sense! Updated

File: firefox-ios/Client/Application/AppLaunchUtil.swift
Comment (linabutler) [inline firefox-ios/Client/Application/AppLaunchUtil.swift:83]:
```suggestion
        var fxaFeatures: RustFxAFeatures = []
```

File: firefox-ios/Client/Application/AppLaunchUtil.swift
Comment (linabutler) [inline firefox-ios/Client/Application/AppLaunchUtil.swift:87]: It looks like we might need to teach `ShareViewController.setupUI()` about this, too—but let's save that for a follow-up!

File: firefox-ios/Client/ApplicationHelper.swift
Comment (linabutler) [inline firefox-ios/Client/ApplicationHelper.swift:53]: Please remember to update this doctoring before landing! 😊

File: firefox-ios/Client/TabManagement/TabManager.swift
Comment (linabutler) [inline firefox-ios/Client/TabManagement/TabManager.swift:86]:
```suggestion
    /// Removes all tabs matching the urls, used when other clients request to close tabs on this device.
```

File: firefox-ios/RustFxA/RustFirefoxAccounts.swift
Comment (linabutler) [inline firefox-ios/RustFxA/RustFirefoxAccounts.swift:181]:
```suggestion
```

I think this comment is redundant now! ✨

File: firefox-ios/Shared/Prefs.swift
Comment (linabutler) [inline firefox-ios/Shared/Prefs.swift:79]: Is this still needed, now that you've switched to using the `FeatureFlagsBoolSetting` and `RustFxAFeatures`?

**Review by linabutler — APPROVED** [review]
:shipit: Great job, @skhamis!

File: firefox-ios/Shared/Prefs.swift
Comment (skhamis) [reply firefox-ios/Shared/Prefs.swift]: Good catch! removed it and just directly used the `.closeRemoteTabs` in the Nimbus case statement.

**Comment by skhamis** [comment]
@mattreaganmozilla I think this is ready to land! Could anyone with landing permissions land this? Thanks!

**Pull request merged**
